builder/amazon/common/step_run_spot_instance.go

```diff
@@ -291,40 +289,39 @@ func (s *StepRunSpotInstance) Run(ctx context.Context, state multistep.StateBag)
 				// we can wait on those operations, this can be removed.
 				return true
 			}
-			return request.IsErrorRetryable(err)
+			return false
 		},
 		RetryDelay: (&retry.Backoff{InitialBackoff: 500 * time.Millisecond, MaxBackoff: 30 * time.Second, Multiplier: 2}).Linear,
 	}.Run(ctx, func(ctx context.Context) error {
 		createOutput, err = ec2conn.CreateFleet(createFleetInput)
-
 		if err == nil && createOutput.Errors != nil {
 			err = fmt.Errorf("errors: %v", createOutput.Errors)
 		}
+		// We can end up with errors because one of the allowed availability
+		// zones doesn't have one of the allowed instance types; as long as
+		// an instance is launched, these errors aren't important.
+		if len(createOutput.Instances) > 0 {
+			if err != nil && !strings.Contains(err.Error(), "Invalid IAM Instance Profile name") {
+				log.Printf("create request failed %v", err)
+				return nil
+			}
+		}
 		if err != nil {
 			log.Printf("create request failed %v", err)
 		}
```

Comment:
Q, would it be a good idea to return nil since an instance was created ?:

```suggestion
			return nil
		}
		if err != nil {
			log.Printf("create request failed %v", err)
		}
```

Reply:
I can remove the `strings.Contains` check but I wanted to log the error anyways so the user will be aware that one or more of the instances have failed

Reply:
is it possible that we ask for multiple instances at once ?

Reply:
yes, I think that's why it's an array of errors because of the error of each instance creation. For the example that I'm using which a spot type is not available at the region, the array contains this error for one and the iam instance profile for the other one.

Reply:
The first error array return for the example I'm using
```
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin: create request failed errors: [{
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidFleetConfiguration",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorMessage: "Your requested instance type ( m5.large) is not supported in your requested Availability Zone (ap-southeast-2c).",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       InstanceType: "m5.large"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     }
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin: } {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidFleetConfiguration",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorMessage: "Your requested instance type ( m5.large) is not supported in your requested Availability Zone (ap-southeast-2a).",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       InstanceType: "m5.large"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     }
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin: } {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidFleetConfiguration",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorMessage: "Your requested instance type ( m5.large) is not supported in your requested Availability Zone (ap-southeast-2b).",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       InstanceType: "m5.large"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     }
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin: } {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidParameterValue",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorMessage: "Value (packer-5f64bb68-c85d-337d-0212-9e404e424ac2) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       InstanceType: "m5.xlarge"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     }
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin: } {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidParameterValue",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorMessage: "Value (packer-5f64bb68-c85d-337d-0212-9e404e424ac2) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       InstanceType: "m5.xlarge"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     }
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin: } {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidParameterValue",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   ErrorMessage: "Value (packer-5f64bb68-c85d-337d-0212-9e404e424ac2) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:       InstanceType: "m5.xlarge"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:     }
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:46 packer-builder-amazon-ebs plugin: }]

```

Reply:
The second one, which don't retry because the other instance was created
```
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin: create request failed for some instances errors: [{
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidFleetConfiguration",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   ErrorMessage: "Your requested instance type ( m5.large) is not supported in your requested Availability Zone (ap-southeast-2c).",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       InstanceType: "m5.large"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     }
    amazon-ebs: Instance ID: i-09f2ed0c30b3f871d
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin: } {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidFleetConfiguration",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   ErrorMessage: "Your requested instance type ( m5.large) is not supported in your requested Availability Zone (ap-southeast-2a).",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       InstanceType: "m5.large"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     }
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin: } {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   ErrorCode: "InvalidFleetConfiguration",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   ErrorMessage: "Your requested instance type ( m5.large) is not supported in your requested Availability Zone (ap-southeast-2b).",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   LaunchTemplateAndOverrides: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     LaunchTemplateSpecification: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       LaunchTemplateId: "lt-0f027471f2862badf",
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       Version: "1"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     },
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     Overrides: {
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:       InstanceType: "m5.large"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:     }
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   },
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin:   Lifecycle: "spot"
2020/09/18 15:51:51 packer-builder-amazon-ebs plugin: }]

```